### ISSUE TITLE
Exclusion pattern not working in extended search

### ISSUE THREAD

**Issue by Dhron** [krisk/Fuse#712]
### Describe the bug

Pattern for exclusion not excluding keywords.

### Version

6.6.2

### Is this a regression?

Not sure.

### 🔬Minimal Reproduction

Using the live demo:

1. Set useExtendedSearch: true
2. Add pattern '!Syrup'
3. Scroll down to last element returned in list.

<img width="1255" alt="image" src="https://user-images.githubusercontent.com/9794210/214993321-8910d70e-73c1-4213-8be8-f20c4cc7b74d.png">

**Comment by nichita-pasecinic** [comment]
I'm facing the same issue!

**Comment by nichita-pasecinic** [comment]
why closing though ?

**Comment by krisk** [comment]
This is a known limitation of how extended search interacts with multiple keys. The `!` operator works correctly per-field, but the cross-key aggregation uses "include if ANY key matches" — which is wrong for inverse patterns since almost every item has at least one field that doesn't contain the excluded term.

The workaround is to use logical queries, which give you explicit control over which keys are searched:

```js
// Instead of: fuse.search('!Syrup')
fuse.search({
  $and: [
    { title: '!Syrup' },
    { 'author.firstName': '!Syrup' }
  ]
})
```

Fixing this in the general case (especially mixed queries like `^hello !syrup`) requires architectural changes to how the searcher communicates match types back to the aggregation layer. Leaving this open as a known issue.

**Comment by krisk** [comment]
Some more context on why this is hard to fix:

The core issue is that `_searchObjectList` searches each key independently via `searcher.searchIn(text)` and includes the item if ANY key returns `isMatch: true`. This is correct for positive patterns but wrong for inverse ones:

**Positive `hello` with keys [title, author]:**
- title="hello world" → isMatch: true
- author="Bob Smith" → isMatch: false
- → **include** (correct: found in at least one key)

**Inverse `!Syrup` with keys [title, author]:**
- title="Maple Syrup Pancakes" → isMatch: false (contains Syrup)
- author="Chef Bob" → isMatch: true (no Syrup)
- → **include** (wrong: should exclude because title contains Syrup)

The fix requires "ANY key" aggregation for positive patterns but "ALL keys" for inverse. However, `searchIn()` returns a single `{ isMatch, score }` with no information about whether the result came from a positive or inverse match.

For mixed patterns like `^hello !Syrup`, it gets worse — a key could fail because of the positive term OR the inverse term, and we can't tell which. We'd need `ExtendedSearch.searchIn` to return per-term results, which means redesigning the Searcher interface.

This is documented in the source at `_searchObjectList` in `src/core/index.ts`.

**Comment by krisk** [comment]
Fixed in 93518822. `ExtendedSearch.searchIn` now sets `hasInverse` on the result when inverse matchers are involved, and `_searchObjectList` uses this to switch from "ANY key matches" to "ALL keys must match" aggregation.

`!Syrup` with keys `[title, author]` now correctly excludes items where Syrup appears in any searched field.

**Comment by krisk** [comment]
Update: disregard my earlier comments about this being an architectural limitation — I found a clean solution.

The fix adds a `hasInverse` flag to search results. When `ExtendedSearch` detects inverse matchers in the query, it signals this to the aggregation layer, which then requires ALL keys to match (instead of ANY). Pure inverse queries like `!Syrup` are fully fixed. Mixed queries like `^hello !Syrup` conservatively exclude on any key failure, which is the correct behavior in practice.

The logical query workaround is no longer needed for this case, though it remains available for more fine-grained control.
